### ISSUE TITLE
Redefine task granularity for parallelism

### ISSUE THREAD

**Issue by kpamnany** [jeff-regier/Celeste.jl#219]
As there is too much variance in the number of sources in different patches of sky, we need a different task granularity for parallelizing. @jeff-regier had suggested we distribute sources instead, which would require some pre-processing? Can we discuss this here?

From the parallelism point of view, we can use sources as tasks, both at the node level and at the thread level. I think this will work well.

**Comment by jeff-regier** [comment]
As I envision it, pre-processing / staging would create a different jld file for each source, containing all the raw data needed to process that source, including the existing catalog entries. (Is that too many files? We could also bundle everything into one HDF5 file.) Whereas the main dataset is 20 TB, the output from pre-processing might be more like 1 or 2 TB, since most of the sky is dark.

The main processing stage would use Dtree, and no other levels of parallelism. Each task operates on one source, and includes both PSF fitting and optimization.

I can see this approach isn't very exciting from an HPC perspective, but it could lead to really good load balancing.

**Comment by kpamnany** [comment]
I don't think we can run efficiently on Cori-II without using threads. Using Dtree alone won't really work I think because then we'd have to start 250 processes per node which would probably be prohibitively expensive in memory terms. However, I don't think that's a problem -- we run a process per socket, use Dtree to distribute sets of sources to processes, and threads to actually optimize those sources. So, we have only one task granularity but use that task at both levels. I think that's a perfectly fine HPC story.

Is there a `read_sdss_field()` equivalent that we can use given an `objid` that will load only the relevant image(s) for that object?

If we eliminate tiles, then is `initialize_model_params()` computationally expensive? What actually would get stored in the JLD file per-source, based on the current steps?

Will the main processing stage do the PSF fitting or is that pre-done?

**Comment by jeff-regier** [comment]
Yes, you can run celeste on a single objid with the subcommand `infer-object` rather than `infer-box`.

With or without tiles, `initialize_model_params()` really shouldn't be computational expensive. The run time should all be almost entirely from 1) load images, 2) fitting the psf, and 3) optimizing.

Yes, the main processing stage should do both psf fitting and optimizing.

**Comment by jeff-regier** [comment]
e.g.

```
./celeste.jl --logging=DEBUG infer-object 4263 5 119 1237663784734490990 $SCRATCH
```

**Comment by jeff-regier** [comment]
Ah, I forgot that `initialize_model_params` evaluates the objective funding once to decide what pixels are influenced by each source. In the future, and especially when we aren't using existing catalogs for initialization, we may want to use a "blob extraction" routine to find bright spots and limit processing to them. But either way, `initialize_model_params` is going to require some constant fraction of the cycles needed to optimize the sources, maybe between 1% and 5℅.

**Comment by jeff-regier** [comment]
PR #224 addresses this issue, more or less.

**Comment by kpamnany** [comment]
If we're to use sources as tasks, we need a way to map a task ID to an object ID. Until we work out how to do that, I think this should remain open.

Also, here's pretty clear evidence as to why we can't use sky areas:

```
[29] processing 23 sources in 200.65, 200.675, 38.225, 38.25
[55] processing 1 sources in 200.45, 200.475, 38.225, 38.25
```

Both are 0.025x0.025 areas of sky. It's pretty hard to balance load with a 23x difference in work item runtime.

There's a lot of variance in source optimization time also.

```
[42] timing: opt_srcs=909.175150801
[42] timing: num_srcs=27
[42] timing: average opt_srcs=33.67315373337037
[41] timing: opt_srcs=773.9899201740001
[41] timing: num_srcs=121
[41] timing: average opt_srcs=6.396610910528926
```

But hopefully this can be balanced effectively.

**Comment by jeff-regier** [comment]
In the future we won't know the number of sources exactly, because we won't have an existing catalog to use for initialization. So I'm wary of a parallelization scheme based on knowing the number of sources ahead of time. (Also, some sources just take much longer than others, as your results seem to show)

The number of bright pixels in each sky area may be a better proxy for run time than the number of sources. The number of bright pixels can be computed pretty quickly, with one or two passes over the pixels in each region.

Perhaps we could continue to parallelize across sky areas, but vary their sizes, based on the number of bright pixels in each region.

**Comment by jeff-regier** [comment]
Closed in #274
